File: backend/src/main/java/ca/metricalsky/winston/entity/fetch/YouTubeRequestEntity.java
package ca.metricalsky.winston.entity.fetch;

import jakarta.persistence.Basic;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.OffsetDateTime;

@Entity
@Table(name = "youtube_requests")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class YouTubeRequestEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Basic(optional = false)
    @Column(name = "fetch_action_id")
    private Long fetchActionId;

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "request_type")
    private RequestType requestType;

    @Basic(optional = false)
    @Column(name = "object_id")
    private String objectId;

    @Column(name = "published_after")
    private String publishedAfter;

    @Column(name = "published_before")
    private String publishedBefore;

    @Column(name = "page_token")
    private String pageToken;

    @Basic(optional = false)
    @Column(name = "requested_at")
    private OffsetDateTime requestedAt;

    @Column(name = "http_status")
    private Integer httpStatus;

    @Column(name = "item_count")
    private Integer itemCount;

    @Column(name = "error")
    private String error;

    @Column(name = "responded_at")
    private OffsetDateTime respondedAt;

    public enum RequestType {
        CHANNELS,
        ACTIVITIES,
        COMMENTS,
        REPLIES,
    }
}

File: database/src/main/kotlin/ca/metricalsky/winston/database/entity/fetch/YouTubeRequestEntity.kt
package ca.metricalsky.winston.database.entity.fetch

import jakarta.persistence.*
import java.time.OffsetDateTime

@Entity
@Table(name = "youtube_requests")
class YouTubeRequestEntity(

    @Basic(optional = false)
    @Column(name = "fetch_action_id")
    var fetchActionId: Long? = null,

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "request_type")
    var requestType: RequestType,

    @Basic(optional = false)
    @Column(name = "object_id")
    var objectId: String,

    @Basic(optional = false)
    @Column(name = "requested_at")
    var requestedAt: OffsetDateTime = OffsetDateTime.now(),
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Long? = null

    @Column(name = "published_after")
    var publishedAfter: String? = null

    @Column(name = "published_before")
    var publishedBefore: String? = null

    @Column(name = "page_token")
    var pageToken: String? = null

    @Column(name = "http_status")
    var httpStatus: Int? = null

    @Column(name = "item_count")
    var itemCount: Int? = null

    @Column(name = "error")
    var error: String? = null

    @Column(name = "responded_at")
    var respondedAt: OffsetDateTime? = null

    enum class RequestType {
        ACTIVITIES,
        CHANNELS,
        COMMENTS,
        PLAYLIST_ITEMS,
        REPLIES,
        VIDEOS,
    }
}
